imprecise types on `urllib3.Retry.new` / `urllib3.Retry.increment`

### Subject

when attempting to subclass `urllib3.Retry` -- one would want to implement `def increment(...) -> Self:` as both `increment` and `new` [preserve the type](https://github.com/urllib3/urllib3/blob/733f638a2faa02b4ff8a9f3b5668949d39396b8b/src/urllib3/util/retry.py#L264) but label the return value as `Retry` (essentially breaking subclassing types): https://github.com/urllib3/urllib3/blob/733f638a2faa02b4ff8a9f3b5668949d39396b8b/src/urllib3/util/retry.py#L243

if the codebase were python 3.11+ I would use `from typing import Self` -- it appears elsewhere in the codebase one-off `SelfT` `TypeVar`s are created -- that pattern could be followed here

alternatively the `TYPE_CHECKING` trick could be used and is understood by the current major type checkers:

```
if TYPE_CHECKING:
    from typing_extensions import Self
```


### Environment

N/A -- latest primary branch source, mypy 1.9


### Steps to Reproduce

```python
from typing import Any, Self
import urllib3
class MyRetry(urllib3.Retry):
    def increment(self, *a: Any, **k: Any) -> Self:  # type error on this line!
        return super().increment(*a, **k)
```

```console
$ mypy t.py
t.py:5: error: Incompatible return value type (got "Retry", expected "Self")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```

### Expected Behavior

no errors

### Actual Behavior

shown above

--- a/changelog/3363.bugfix.rst
+++ b/changelog/3363.bugfix.rst
@@ -1,0 +1,1 @@
+Consistently use ``typing.Self`` for return types representing copying actions.

--- a/changelog/3363.bugfix.rst
+++ b/changelog/3363.bugfix.rst
@@ -1,1 +1,1 @@
-Consistently use ``typing.Self`` for return types representing copying actions.
+Consistently used ``typing.Self`` for return types representing copying actions.

--- a/src/urllib3/_collections.py
+++ b/src/urllib3/_collections.py
@@ -427,7 +427,7 @@ class HTTPHeaderDict(typing.MutableMapping[str, str]):
             val = other.getlist(key)
             self._container[key.lower()] = [key, *val]
 
-    def copy(self) -> HTTPHeaderDict:
+    def copy(self) -> Self:
         clone = type(self)()
         clone._copy_from(self)
         return clone
@@ -462,7 +462,7 @@ class HTTPHeaderDict(typing.MutableMapping[str, str]):
         self.extend(maybe_constructable)
         return self
 
-    def __or__(self, other: object) -> HTTPHeaderDict:
+    def __or__(self, other: object) -> Self:
         # Supports merging header dicts using operator |
         # combining items with add instead of __setitem__
         maybe_constructable = ensure_can_construct_http_header_dict(other)
@@ -472,7 +472,7 @@ class HTTPHeaderDict(typing.MutableMapping[str, str]):
         result.extend(maybe_constructable)
         return result
 
-    def __ror__(self, other: object) -> HTTPHeaderDict:
+    def __ror__(self, other: object) -> Self:
         # Supports merging header dicts using operator | when other is on left side
         # combining items with add instead of __setitem__
         maybe_constructable = ensure_can_construct_http_header_dict(other)

--- a/src/urllib3/connectionpool.py
+++ b/src/urllib3/connectionpool.py
@@ -55,13 +55,13 @@ if typing.TYPE_CHECKING:
     import ssl
     from typing import Literal
 
+    from typing_extensions import Self
+
     from ._base_connection import BaseHTTPConnection, BaseHTTPSConnection
 
 log = logging.getLogger(__name__)
 
 _TYPE_TIMEOUT = typing.Union[Timeout, float, _TYPE_DEFAULT, None]
-
-_SelfT = typing.TypeVar("_SelfT")
 
 
 # Pool objects
@@ -95,7 +95,7 @@ class ConnectionPool:
     def __str__(self) -> str:
         return f"{type(self).__name__}(host={self.host!r}, port={self.port!r})"
 
-    def __enter__(self: _SelfT) -> _SelfT:
+    def __enter__(self) -> Self:
         return self
 
     def __exit__(

--- a/src/urllib3/poolmanager.py
+++ b/src/urllib3/poolmanager.py
@@ -28,6 +28,8 @@ if typing.TYPE_CHECKING:
     import ssl
     from typing import Literal
 
+    from typing_extensions import Self
+
 __all__ = ["PoolManager", "ProxyManager", "proxy_from_url"]
 
 
@@ -50,8 +52,6 @@ SSL_KEYWORDS = (
 # Default value for `blocksize` - a new parameter introduced to
 # http.client.HTTPConnection & http.client.HTTPSConnection in Python 3.7
 _DEFAULT_BLOCKSIZE = 16384
-
-_SelfT = typing.TypeVar("_SelfT")
 
 
 class PoolKey(typing.NamedTuple):
@@ -214,7 +214,7 @@ class PoolManager(RequestMethods):
         self.pool_classes_by_scheme = pool_classes_by_scheme
         self.key_fn_by_scheme = key_fn_by_scheme.copy()
 
-    def __enter__(self: _SelfT) -> _SelfT:
+    def __enter__(self) -> Self:
         return self
 
     def __exit__(

--- a/src/urllib3/util/retry.py
+++ b/src/urllib3/util/retry.py
@@ -21,6 +21,8 @@ from ..exceptions import (
 from .util import reraise
 
 if typing.TYPE_CHECKING:
+    from typing_extensions import Self
+
     from ..connectionpool import ConnectionPool
     from ..response import BaseHTTPResponse
 
@@ -240,7 +242,7 @@ class Retry:
         )
         self.backoff_jitter = backoff_jitter
 
-    def new(self, **kw: typing.Any) -> Retry:
+    def new(self, **kw: typing.Any) -> Self:
         params = dict(
             total=self.total,
             connect=self.connect,
@@ -429,7 +431,7 @@ class Retry:
         error: Exception | None = None,
         _pool: ConnectionPool | None = None,
         _stacktrace: TracebackType | None = None,
-    ) -> Retry:
+    ) -> Self:
         """Return a new Retry object with incremented retry counters.
 
         :param response: A response object, or None, if the server did not

--- a/src/urllib3/util/ssltransport.py
+++ b/src/urllib3/util/ssltransport.py
@@ -10,10 +10,11 @@ from ..exceptions import ProxySchemeUnsupported
 if typing.TYPE_CHECKING:
     from typing import Literal
 
+    from typing_extensions import Self
+
     from .ssl_ import _TYPE_PEER_CERT_RET, _TYPE_PEER_CERT_RET_DICT
 
 
-_SelfT = typing.TypeVar("_SelfT", bound="SSLTransport")
 _WriteBuffer = typing.Union[bytearray, memoryview]
 _ReturnValue = typing.TypeVar("_ReturnValue")
 
@@ -70,7 +71,7 @@ class SSLTransport:
         # Perform initial handshake.
         self._ssl_io_loop(self.sslobj.do_handshake)
 
-    def __enter__(self: _SelfT) -> _SelfT:
+    def __enter__(self) -> Self:
         return self
 
     def __exit__(self, *_: typing.Any) -> None:

